money-aspectj/src/main/scala/com/comcast/money/aspectj/TraceAspect.scala

```diff
@@ -39,22 +42,70 @@ class TraceAspect extends Reflections with TraceLogging {
 
   @Around("traced(traceAnnotation)")
   def adviseMethodsWithTracing(joinPoint: ProceedingJoinPoint, traceAnnotation: Traced): AnyRef = {
-    val key: String = traceAnnotation.value
-    var result = true
+    if (traceAnnotation.async()) {
+      traceMethodAsync(joinPoint, traceAnnotation)
+    } else {
+      traceMethod(joinPoint, traceAnnotation)
+    }
+  }
+
+  private def traceMethod(joinPoint: ProceedingJoinPoint, traceAnnotation: Traced): AnyRef = {
+    val key = traceAnnotation.value()
     val oldSpanName = mdcSupport.getSpanNameMDC
+    var result = true
+
     try {
       tracer.startSpan(key)
       mdcSupport.setSpanNameMDC(Some(key))
       traceMethodArguments(joinPoint)
-      joinPoint.proceed
+
+      joinPoint.proceed()
     } catch {
       case t: Throwable =>
-        result = if (exceptionMatches(t, traceAnnotation.ignoredExceptions())) true else false
+        result = exceptionMatches(t, traceAnnotation.ignoredExceptions())
         logException(t)
         throw t
     } finally {
+      tracer.stopSpan(result)
       mdcSupport.setSpanNameMDC(oldSpanName)
+    }
+  }
+
+  private def traceMethodAsync(joinPoint: ProceedingJoinPoint, traceAnnotation: Traced): AnyRef = {
+    val key = traceAnnotation.value()
+    val asyncKey = key + "-async"
```

Comment:
why do we append -async to the name?

Reply:
I'm creating two spans.  I'm giving the second span which represents the async task the suffix `-async`.  I could change that to do something else, or make it configurable.

Reply:
Ok, that is where I was confused.  I don't understand why we need two separate spans.

We use Money heavily with scala futures, and all tracing is declaritive in the code.  We want to give the user the ability to start / stop however we want.

It is a little unclear here with the `@Traced` annotation as to when the span actually starts.  The way we have done this is that the span starts when the method is invoked and completes when the Future (Promise) ends.  Something conceptually like this...

```
def doSomethingTraced(foo: String): Future[String] = {
  val span = startSpan("myName")
  val fut = doSomethingInTheFuture(foo)
  fut.onComplete { case Success(x) => span.stop(true); case Failure(e) => span.stop(false) }
}
```

Reply:
I see.  That is the approach I was taking in the first iteration and I have no problem going back to that model.  If I do that then the code above could be reduced to:

```scala
@Traced(value = "myName", async = true)
def doSomethingTraced(foo: String): Future[String] =
    doSomethingInTheFuture(foo)
```